BIP 350: Implement Bech32m and use it for v1+ segwit addresses (0.20 backport)

sipa opened the pull request:
Backport of #20861 for 0.20. Also includes #21471.

sipa committed:
Implement Bech32m encoding/decoding

Github-Pull: #20861
Rebased-From: da2bb6976dadeec682d163c258c9afecc87d6428

sipa committed:
Add Bech32m test vectors

Github-Pull: #20861
Rebased-From: 25b1c6e13ddf1626210d5e3d37298d1f3a78a94f

sipa committed:
Use Bech32m encoding for v1+ segwit addresses

This also includes updates to the Python test framework implementation,
test vectors, and release notes.

Github-Pull: #20861
Rebased-From: fe5e495c31de47b0ec732b943db11fe345d874af

fjahr committed:
naming nits

Github-Pull: #20861
Rebased-From: 03346022d611871f2cc185440b19d928b9264d9d

jnewbery commented on doc/bips.md:
```diff
@@ -42,3 +42,4 @@ BIPs that are implemented by Bitcoin Core (up-to-date up to **v0.19.0**):
 * [`BIP 173`](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki): Bech32 addresses for native Segregated Witness outputs are supported as of **v0.16.0** ([PR 11167](https://github.com/bitcoin/bitcoin/pull/11167)). Bech32 addresses are generated by default as of **v0.20.0** ([PR 16884](https://github.com/bitcoin/bitcoin/pull/16884)).
 * [`BIP 174`](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki): RPCs to operate on Partially Signed Bitcoin Transactions (PSBT) are present as of **v0.17.0** ([PR 13557](https://github.com/bitcoin/bitcoin/pull/13557)).
 * [`BIP 176`](https://github.com/bitcoin/bips/blob/master/bip-0176.mediawiki): Bits Denomination [QT only] is supported as of **v0.16.0** ([PR 12035](https://github.com/bitcoin/bitcoin/pull/12035)).
+* [`BIP 350`](https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki): Addresses for native v1+ segregated Witness outputs use Bech32m instead of Bech32 as of **v22.0** ([PR 20861](https://github.com/bitcoin/bitcoin/pull/20861)).
```
```suggestion
* [`BIP 350`](https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki): Addresses for native v1+ segregated Witness outputs use Bech32m instead of Bech32 as of **v0.20.2** ([PR 20861](https://github.com/bitcoin/bitcoin/pull/20861)).
```

Also update top of document to:

`BIPs that are implemented by Bitcoin Core (up-to-date up to **v0.20.2**):`

benthecarman replied:
This is backported to v0.19.X actually

#21472

jnewbery replied:
Perhaps, although I think this should document the earliest release that included the BIP *in this branch*. It doesn't really make sense to say which of 0.19.n or 0.20.m is earlier, since they're parallel release tracks. In #21469, the first release is 0.21.1, and in #20861, the first release is 22.0.

I also don't think that a 0.19 backport is necessary, but others may disagree.

sipa replied:
Listed v0.20.2.

jnewbery reviewed (commented):
One documentation comment. Otherwise, this is exactly the same as I got when I cherry-picked these commits to the 0.20 branch.

sipa committed:
Backport invalid address tests

Reduced version of the test from master/#20861 by John Newbery.

Github-Pull: #20861
Rebased-From: fe5e495c31de47b0ec732b943db11fe345d874af

sipa commented:
Anyone have a clue why appveyor fails here?

Also, don't we need cirrus CI changes backported to 0.20 first?

MarcoFalke commented:
The msvc is now using a dependency file (https://github.com/bitcoin/bitcoin/blob/0.21/build_msvc/vcpkg.json), which might need to be backported to fix the appveyor ci.

Looks like travis isn't fully shut down yet and still runs on the branch. This is fine, because the ci system uses bash scripts that don't care which infrastructure they run on.

jnewbery commented:
utACK c0f85fd8503a6e6f66cb8659bd88deec92b96eec

Verified the range-diff. Only difference is updating bips.md and adding the rpc_invalid_address_message.py test.

maflcko reviewed (commented):
range-diff-only ACK c0f85fd850 🐔

Compared against the version merged into the 0.21 branch with git range-diff. Only changes:

* Account for naming difference in tests: decode_segwit_address vs decode
* Version number in doc/bips.md
* Remove signet mention from release notes snippet
* Other minor merge conflicts in tests
* Drop signet commit 7dfe406e20


<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

range-diff-only ACK c0f85fd850 🐔

Compared against the version merged into the 0.21 branch with git range-diff. Only changes:

* Account for naming difference in tests: decode_segwit_address vs decode
* Version number in doc/bips.md
* Remove signet mention from release notes snippet
* Other minor merge conflicts in tests
* Drop signet commit 7dfe406e20

-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhYfwv+LPVF9v+957bB2YPLecWl1tfq4Q/rvj4+r8neaXzEdyyMmS3norOIxhKE
f7WhOInn87DnDroCNfJjss/Jz6fMkKKEjeJ0N8lYwva28TqGhhQnWQyharME9SEw
2/+KXAhg7Y8CUwfx1V0yd+wZhZhy9c4+rcc4yeIte/CyPpZuX2ch+bOuxd8zaA9a
vKv8NNCeb17NgMLaDCrOkt3kVrVzmTR/4z82eLlHoo6BKMVqZNs1Kof51vpe3WqI
7GNZzOZ2vwEPYHdrjCMHu4/GX/LZ50luAhZWr9dQSokb2q73yM2F8z7AcUkgP4sp
nZhUKJTI+VQXmTcVajaf/7+uhlm0KnSJM9kn3q/pgQlzwtPNiR+XBvfjnUpht9db
d/VfTegviM8tpEpfxyMys0mY6GS9aFxdNCez0XWeX2bIsbzMB8oEyTb+LF+oAwbk
rLPmYU10BrVbpMxPHja7rbhVGejL5ShwLJe+olRXMsr/mxhO4XeVF+RiiUXAzbXO
vZgTSNND
=mCLL
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `a9384816e3019705b939aef50969fdb1f644619973c3523f30929fc5022c81ee  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108a9384816e3019705b939aef50969fdb1f644619973c3523f30929fc5022c81eef010e6249782d0ee05da302c38de63c7adeb08fff010284d1a72162edef37f705594ce08b1cb08f020ee26c53816e8d50522c7f652e05953d175f184ffa489deeb40a69aa08b76739308f020ddd8fcb9200f5c20e16a15375f66f777e22b82d7a741dd3418d57a9a672cc5a108f1046065a780f008f72c225bfc8223de0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0105f4e72576b0995660daa1500cef9a68808f104606599a8f00846d17e998af5121d0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010670d0d4e3de8fb78c0c3210d7f8ad41108f0203d917d68d3d93dc5d268bc11fd376bfb3b8ebd65b65bb007cfe0bfed7316108708f020f259b3efb8d4f03d2bff74a9b0bea94d04906b74181c25f5df204ed25a1ca2a308f1046065a780f0084d45fa5fcb9c8bcd0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010be950b178f4e0cfa13f0c4423a809e9808f1046065a780f0086e3a565c94edfdcc0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>
